esp-hal/src/gpio/mod.rs

```diff
@@ -1907,65 +1962,86 @@ where
     /// Create GPIO input driver for a [Pin] with the provided [Pull]
     /// configuration.
     #[inline]
-    pub fn new(pin: impl crate::peripheral::Peripheral<P = P> + 'd, pull: Pull) -> Self {
-        crate::into_ref!(pin);
-        pin.init_input(pull == Pull::Down, pull == Pull::Up, private::Internal);
+    pub fn new_typed(pin: impl crate::peripheral::Peripheral<P = P> + 'd, pull: Pull) -> Self {
+        let pin = Flex::new_typed(pin);
+
+        Self::new_inner(pin, pull)
+    }
+
+    fn new_inner(mut pin: Flex<'d, P>, pull: Pull) -> Self {
+        pin.set_as_input(pull);
+
         Self { pin }
     }
 
     /// Get whether the pin input level is high.
     #[inline]
     pub fn is_high(&self) -> bool {
-        self.pin.is_input_high(private::Internal)
+        self.get_level() == Level::High
     }
 
     /// Get whether the pin input level is low.
     #[inline]
     pub fn is_low(&self) -> bool {
-        !self.is_high()
+        self.get_level() == Level::Low
     }
 
     /// Get the current pin input level.
     #[inline]
     pub fn get_level(&self) -> Level {
-        self.is_high().into()
+        self.pin.get_level()
     }
 
     /// Listen for interrupts
     #[inline]
     pub fn listen(&mut self, event: Event) {
-        self.pin.listen(event, private::Internal);
+        self.pin.listen(event);
     }
 
     /// Stop listening for interrupts
     pub fn unlisten(&mut self) {
-        self.pin.unlisten(private::Internal);
+        self.pin.unlisten();
     }
 
     /// Clear the interrupt status bit for this Pin
     #[inline]
     pub fn clear_interrupt(&mut self) {
-        self.pin.clear_interrupt(private::Internal);
+        self.pin.clear_interrupt();
     }
 
     /// Checks if the interrupt status bit for this Pin is set
     #[inline]
     pub fn is_interrupt_set(&self) -> bool {
-        self.pin.is_interrupt_set(private::Internal)
+        self.pin.is_interrupt_set()
     }
 
     /// Enable as a wake-up source.
     ///
     /// This will unlisten for interrupts
     #[inline]
     pub fn wakeup_enable(&mut self, enable: bool, event: WakeEvent) {
-        self.pin.wakeup_enable(enable, event, private::Internal);
+        self.pin.wakeup_enable(enable, event);
     }
 }
 
 /// GPIO open-drain output driver.
-pub struct OutputOpenDrain<'d, P> {
-    pin: PeripheralRef<'d, P>,
+pub struct OutputOpenDrain<'d, P = ErasedPin> {
```

Comment:
Should this be `P = AnyPin<'d>`, we have `ErasedPin` as `#[doc(hidden)]`, and this somewhat leaks this. If it's possible, I'd like `ErasedPin` to be completely private tbh

Reply:
I guess I struggle to see the difference between `ErasedPin` and `AnyPin` other than with `AnyPin` you can specify an inversion - maybe `AnyPin` should just be a light wrapper to hide the details of `ErasedPin` and an inverted pin should be a new type?

Reply:
That is a very silly idea, there's no reason lugging the `is_inverted` field around because it's not used for GPIOs. I want to remove the idea of our current AnyPin, and replace it with something else for the peripheral connections.

Reply:
You were clear enough, you just pushed your second message before I responded to the first one.

Unfortunately that ugly enum is a very convenient way to abstract over the two banks :D Whether we want to wrap this with a newtype... IDK I'm sick of all the delegates here, but why not.

Reply:
Will be addressed in an upcoming PR, if that's all right with you.